File: typescript/src/csv.ts
// 一覧・帳票の CSV 出力。列 + 行 → 文字列。Dart / Java 版と同結果。
//
// 既定では列の format を通した表記で書く（画面と同じ見た目）。Excel で計算
// させたいときは raw: true。文字コード変換は Framework の外＝出力先の責務。

import { type ColumnDefinition } from "./definition.js";
import { FormatterRegistry } from "./formatter.js";

/** CSV の書き方。既定は Excel（日本語 Windows）で開くのが楽な組み合わせ。 */
export interface CsvOptions {
  /** 見出し行（列ラベル）を出すか。 */
  header: boolean;
  /** 区切り文字。 */
  delimiter: string;
  /** 改行: `crlf`（既定）か `lf`。 */
  newline: string;
  /** 先頭に BOM を付けるか（Excel の文字化け対策）。 */
  bom: boolean;
  /** format を通さず生の値を書くか。 */
  raw: boolean;
  /**
   * 出力先に渡す文字コードの名前（既定 `utf-8`）。**変換はここではしない**
   * ＝Framework は文字列までで、バイト列にするのは出力先の責務。
   */
  charset: string;
}

/** 既定の文字コード名。 */
export const UTF8_CHARSET = "utf-8";

export const defaultCsvOptions: CsvOptions = {
  header: true,
  delimiter: ",",
  newline: "crlf",
  bom: false,
  raw: false,
  charset: UTF8_CHARSET,
};

/** DSL の config（export アクション）から読む。 */
export function csvOptionsFromConfig(
  config: Record<string, unknown>,
): CsvOptions {
  return {
    header: typeof config.header === "boolean" ? config.header : true,
    delimiter: config.delimiter == null ? "," : String(config.delimiter),
    newline: config.newline == null ? "crlf" : String(config.newline),
    bom: config.bom === true,
    raw: config.raw === true,
    charset:
      config.charset == null ? UTF8_CHARSET : String(config.charset),
  };
}

/** UTF-8 か（表記ゆれを吸収する）。 */
export const isUtf8Charset = (charset: string): boolean =>
  ["utf-8", "utf8", "utf_8"].includes(charset.toLowerCase());

/**
 * 実際に BOM を付けるか。宣言されていて、かつ UTF-8 のときだけ
 * （Shift_JIS などに BOM は無く、付けると先頭のセルにゴミが3バイト入る）。
 */
export const wantsBom = (options: CsvOptions): boolean =>
  options.bom && isUtf8Charset(options.charset);

/**
 * rows を columns の順に CSV へ書き出す。列が無ければ空文字。
 * 区切り・引用符・改行を含む値は引用し、引用符は2つに重ねる（RFC 4180）。
 */
export function toCsv(
  columns: ColumnDefinition[],
  rows: Record<string, unknown>[],
  options: CsvOptions = defaultCsvOptions,
  formatters: FormatterRegistry = new FormatterRegistry(),
): string {
  if (columns.length === 0) return "";
  const lineBreak = options.newline === "lf" ? "\n" : "\r\n";
  const lines: string[] = [];

  if (options.header) {
    lines.push(
      columns.map((c) => escape(c.label, options.delimiter)).join(options.delimiter),
    );
  }
  for (const row of rows) {
    lines.push(
      columns
        .map((c) => escape(cell(c, row[c.field], options, formatters), options.delimiter))
        .join(options.delimiter),
    );
  }
  const bom = wantsBom(options) ? "\ufeff" : "";
  return bom + lines.map((line) => line + lineBreak).join("");
}

function cell(
  column: ColumnDefinition,
  value: unknown,
  options: CsvOptions,
  formatters: FormatterRegistry,
): string {
  if (!options.raw && column.format) {
    return formatters.format(column.format, value, column.config);
  }
  return value == null ? "" : String(value);
}

/** 引用が要るのは区切り・引用符・改行を含むときだけ。 */
function escape(value: string, delimiter: string): string {
  const needsQuotes =
    value.includes(delimiter) ||
    value.includes('"') ||
    value.includes("\n") ||
    value.includes("\r");
  return needsQuotes ? `"${value.replaceAll('"', '""')}"` : value;
}

File: java/src/main/java/io/hatake/core/Csv.java
package io.hatake.core;

import java.util.List;
import java.util.Map;

/**
 * 一覧・帳票の CSV 出力。列 + 行 → 文字列。Dart / TypeScript 版と同結果。
 *
 * <p>既定では列の {@code format} を通した表記で書く（画面と同じ見た目）。Excel で
 * 計算させたいときは {@code raw}。文字コード変換（Shift_JIS 等）は Framework の
 * 外＝出力先の責務。
 */
public final class Csv {

    private Csv() {
    }

    /**
     * CSV の書き方。既定は Excel（日本語 Windows）で開くのが楽な組み合わせ。
     *
     * @param header 見出し行（列ラベル）を出すか
     * @param delimiter 区切り文字
     * @param newline 改行（{@code crlf} / {@code lf}）
     * @param bom 先頭に BOM を付けるか（Excel の文字化け対策。UTF-8 のときだけ効く）
     * @param raw {@code format} を通さず生の値を書くか
     * @param charset 出力先に渡す文字コードの名前（既定 {@code utf-8}）。
     *     <b>変換はここではしない</b>＝文字列を作るまでが Framework の仕事で、
     *     バイト列にするのは出力先の責務
     */
    public record Options(
            boolean header,
            String delimiter,
            String newline,
            boolean bom,
            boolean raw,
            String charset) {

        /** 既定の文字コード名。 */
        public static final String UTF8 = "utf-8";

        public static final Options DEFAULT =
                new Options(true, ",", "crlf", false, false, UTF8);

        /** DSL の {@code config}（{@code export} アクション）から読む。 */
        public static Options fromConfig(Map<String, Object> config) {
            if (config == null) {
                return DEFAULT;
            }
            return new Options(
                    !(config.get("header") instanceof Boolean h) || h,
                    config.get("delimiter") == null ? "," : config.get("delimiter").toString(),
                    config.get("newline") == null ? "crlf" : config.get("newline").toString(),
                    Boolean.TRUE.equals(config.get("bom")),
                    Boolean.TRUE.equals(config.get("raw")),
                    config.get("charset") == null ? UTF8 : config.get("charset").toString());
        }

        public String lineBreak() {
            return "lf".equals(newline) ? "\n" : "\r\n";
        }

        /** UTF-8 か（表記ゆれを吸収する）。 */
        public boolean isUtf8() {
            String lower = charset.toLowerCase(java.util.Locale.ROOT);
            return lower.equals("utf-8") || lower.equals("utf8") || lower.equals("utf_8");
        }

        /**
         * 実際に BOM を付けるか。宣言されていて、かつ UTF-8 のときだけ
         * （Shift_JIS などに BOM は無く、付けると先頭のセルにゴミが3バイト入る）。
         */
        public boolean wantsBom() {
            return bom && isUtf8();
        }
    }

    public static String toCsv(
            List<ColumnDefinition> columns, List<Map<String, Object>> rows) {
        return toCsv(columns, rows, Options.DEFAULT, new FormatterRegistry());
    }

    /**
     * {@code rows} を {@code columns} の順に CSV へ書き出す。列が無ければ空文字。
     * 区切り・引用符・改行を含む値は引用し、引用符は2つに重ねる（RFC 4180）。
     */
    public static String toCsv(
            List<ColumnDefinition> columns,
            List<Map<String, Object>> rows,
            Options options,
            FormatterRegistry formatters) {
        if (columns.isEmpty()) {
            return "";
        }
        StringBuilder out = new StringBuilder();
        if (options.wantsBom()) {
            out.append('\uFEFF');
        }
        if (options.header()) {
            for (int i = 0; i < columns.size(); i++) {
                if (i > 0) {
                    out.append(options.delimiter());
                }
                out.append(escape(columns.get(i).label(), options.delimiter()));
            }
            out.append(options.lineBreak());
        }
        for (Map<String, Object> row : rows) {
            for (int i = 0; i < columns.size(); i++) {
                if (i > 0) {
                    out.append(options.delimiter());
                }
                ColumnDefinition column = columns.get(i);
                out.append(escape(
                        cell(column, row.get(column.field()), options, formatters),
                        options.delimiter()));
            }
            out.append(options.lineBreak());
        }
        return out.toString();
    }

    private static String cell(
            ColumnDefinition column,
            Object value,
            Options options,
            FormatterRegistry formatters) {
        if (!options.raw() && column.format() != null) {
            return formatters.format(column.format(), value, column.config());
        }
        return value == null ? "" : value.toString();
    }

    /** 引用が要るのは区切り・引用符・改行を含むときだけ。 */
    private static String escape(String value, String delimiter) {
        boolean needsQuotes = value.contains(delimiter)
                || value.contains("\"")
                || value.contains("\n")
                || value.contains("\r");
        return needsQuotes ? '"' + value.replace("\"", "\"\"") + '"' : value;
    }
}
